src/stan_math_backend/Stan_math_code_gen.ml

```diff
@@ -472,59 +472,104 @@ using stan::math::lgamma;
 using stan::model::prob_grad;
 using namespace stan::math; |}
 
-let rec xform_readdata sizes s =
-  let get_single_size = function
-    | Single e -> e
-    | x ->
-        raise_s
-          [%message
-            "expecting ReadData to just have Single indices"
-              (x : mtype_loc_ad with_expr index)]
+let rec expr_contains_fn fname accum e =
+  match e.expr with
+  | FunApp (_, name, _) when name = fname -> true
+  | x -> fold_expr (expr_contains_fn fname) accum x
+
+let%test "expr contains fn" =
+  internal_funapp FnReadData [] ()
+  |> expr_contains_fn (string_of_internal_fn FnReadData) false
+
+let rec contains_fn fname accum {stmt; _} =
+  fold_statement (expr_contains_fn fname)
+    (fun accum s ->
+      fold_statement (expr_contains_fn fname) (contains_fn fname) accum s.stmt
+      )
+    accum stmt
+
+let fake_stmt stmt = {stmt; smeta= no_span}
+
+let fake_for emeta body =
+  let ten = {expr= Lit (Int, "10"); emeta} in
+  For {loopvar= "lv"; lower= ten; upper= ten; body= fake_stmt (Block [body])}
+  |> fake_stmt
+
+let%test "contains fn" =
+  let f =
+    fake_for ()
+      (fake_for ()
+         (fake_stmt (Assignment (("v", []), internal_funapp FnReadData [] ()))))
   in
-  let get_index_sizes idcs = List.map ~f:get_single_size idcs in
-  match s.stmt with
-  | Assignment
-      ( lhs
-      , { expr=
-            Indexed
-              ( ({expr= FunApp (CompilerInternal, f, _); emeta} as fnapp)
-              , idc1 :: idc2 :: idcs ); _ } )
+  contains_fn
+    (string_of_internal_fn FnReadData)
+    false
+    (fake_stmt (Block [f; fake_stmt Break]))
+
+let pos = "pos__"
+let mir_int i = {expr= Lit (Int, string_of_int i); emeta= internal_meta}
+
+let add_pos_reset ({stmt; smeta} as s) =
+  match stmt with
+  | For {body; _}
+    when contains_fn (string_of_internal_fn FnReadData) false body ->
+      [{stmt= Assignment ((pos, []), loop_bottom); smeta}; s]
+  | _ -> [s]
+
+let rec use_pos_in_readdata {stmt; smeta} =
+  match stmt with
+  | Block
```

Comment:
This is so ugly! Not sure what to do about it.

Reply:
I'm not sure there is anything that _can_ be done; in F# & Haskell there is a neat feature called [Active Patterns](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/active-patterns) and [Pattern Synonyms](https://gitlab.haskell.org/ghc/ghc/wikis/pattern-synonyms) respectively that may be helpful but I think OCaml has no equivalent.

Reply:
Yep.. I know of no good alternative.